### ISSUE TITLE
[REVIEW]: turtleFSI: A Robust and Monolithic FEniCS-based Fluid-Structure Interaction Solver

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2089]
**Submitting author:** @albansouche (<a href="http://orcid.org/0000-0001-7547-7979">Alban Souche</a>)
**Repository:** <a href="https://github.com/KVSlab/turtleFSI.git" target ="_blank">https://github.com/KVSlab/turtleFSI.git</a>
**Version:** v1.4
**Editor:** @Kevin-Mattheus-Moerman
**Reviewers:** @chennachaos, @JaroslavHron
**Archive:** <a href="https://doi.org/10.5281/zenodo.3895443" target="_blank">10.5281/zenodo.3895443</a>

## Status

[![status](https://joss.theoj.org/papers/b7febdaa2709205d40b51227091c3b0b/status.svg)](https://joss.theoj.org/papers/b7febdaa2709205d40b51227091c3b0b)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/b7febdaa2709205d40b51227091c3b0b"><img src="https://joss.theoj.org/papers/b7febdaa2709205d40b51227091c3b0b/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/b7febdaa2709205d40b51227091c3b0b/status.svg)](https://joss.theoj.org/papers/b7febdaa2709205d40b51227091c3b0b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/KVSlab/turtleFSI.git" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@nathanquinlan & @chennachaos & @JaroslavHron, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @Kevin-Mattheus-Moerman know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/KVSlab/turtleFSI.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@albansouche) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/KVSlab/turtleFSI.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@albansouche) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @nathanquinlan, @chennachaos, @JaroslavHron it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1146/annurev.fluid.30.1.539 is OK
- 10.1023/A:1020843529530 is OK
- 10.1016/S0045-7825(00)00381-9 is OK
- 10.1016/j.compstruc.2011.02.019 is OK
- 10.1007/978-3-642-23099-8 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@nathanquinlan, @chennachaos, @JaroslavHron, this is where the review takes place. 

The instructions and checkboxes at the top will guide you through the review. As a reviewer  you can leave comments on this work here. For larger issues we encourage to open an issue on the [project repository](https://github.com/KVSlab/turtleFSI.git) and to link to it here in a comment. 

Let me know if you have questions. Thanks again for your help! :rocket: :robot:

**Comment by Kevin-Mattheus-Moerman** [comment]
@nathanquinlan, @chennachaos, @JaroslavHron, thanks for your help with this review! Can you provide an update on review progress? Thanks!

**Comment by chennachaos** [comment]
Hi @Kevin-Mattheus-Moerman. Apologies for the delay.
I will try to complete the review by the end of this month.

**Comment by arfon** [comment]
Dear authors and reviewers

We wanted to notify you that in light of the current COVID-19 pandemic, JOSS has decided to suspend submission of new manuscripts and to handle existing manuscripts (such as this one) on a "best efforts basis". We understand that you may need to attend to more pressing issues than completing a review or updating a repository in response to a review. If this is the case, a quick note indicating that you need to put a "pause" on your involvement with a review would be appreciated but is not required.

Thanks in advance for your understanding.

_Arfon Smith, Editor in Chief, on behalf of the JOSS editorial team._

**Comment by chennachaos** [comment]
Hi @albansouche
I must say that it is very refreshing to see an open-source software library for FSI problems. 

I have started the review. I have been going through the code and the associated documentation. I am having difficulties in understanding some parts. I would like to understand them at length before providing my detailed remarks.

1.) How, and on what level, does turtleFSI interact with FEniCS?

2.) What is the example problem used in the tests?

3.) Can the solver cope with mismatching meshes at the fluid-solid interface?

4.) Do you have any data on the convergence of Newton-Raphson iterations and how this convergence is affected under different settings of parameters passed to the *newtonsolver*?

5.) a) How is the Jacobian matrix for the monolithic scheme computed?
     b) How is the matrix system resulting from the monolithic scheme solved?

These (5a and 5b) are the two most time-consuming parts in an FSI solver. To gain insights into the computational aspects, I have looked at the graphs you have nicely prepared in the solver performance section. From the graphs, I notice that *one assembly of the Jacobian matrix requires about two orders of magnitude more CPU time than that of one matrix solver*. My observation seems to be in line with your statement "Notice that most of the compute time is spent on the Jacobian operations" in the caption.

This behaviour of the solver, however, is counter-intuitive to me because the matrix solver should be the most expensive part in any implicit scheme. Please clarify why the formation of the Jacobian matrix is so much more expensive.

**Comment by albansouche** [comment]
Hi @chennachaos 
Thank you for your first feedbacks. Here are my replies. 

_1.) How, and on what level, does turtleFSI interact with FeniCS?_
turtleFSI interacts with FEniCS through the python interface. It provides a compact code relatively easy to modify for further development for anyone with basic knowledge of the FEniCS library.

_2.) What is the example problem used in the tests?_
We check all the examples (Turek flag with only solid, only fluid, and FSI) and all the mesh lifting operators in the tests.

_3.) Can the solver cope with mismatching meshes at the fluid-solid interface?_
Unfortunately not. turtleFSI solver only copes with conforming solid and fluid meshes at the interface.

_4.) Do you have any data on the convergence of Newton-Raphson iterations and how this convergence is affected under different settings of parameters passed to the newtonsolver?_
We didn’t document this aspect in details since the convergence rate can vary greatly depending on the problem being solved, the mesh resolution, or the time step…   
In brief, the convergence is almost systematical quadratic when the Jacobian matrix is re-evaluated at each iteration. However, for efficiency reason, most of the time we prefer to reuse the Jacobian matrix over the Newton iterations even if this results in a sub-linear convergence rate. Recomputing the Jacobian matrix is not excessively time consuming by itself, but will  systematically require a new factorization step within the direct solver (here, MUMPS) which is computationally very expensive.

_5.) a) How is the Jacobian matrix for the monolithic scheme computed?_
The Jacobian matrix is numerically evaluated within FEniCS.

_b) How is the matrix system resulting from the monolithic scheme solved?_
It is solved with MUMPS solver available though the FEniCS PETSc backend. It appears to be the most robust solver available from FEniCS for FSI problems.

_These (5a and 5b) are the two most time-consuming parts in an FSI solver. To gain insights into the computational aspects, I have looked at the graphs you have nicely prepared in the solver performance section. From the graphs, I notice that one assembly of the Jacobian matrix requires about two orders of magnitude more CPU time than that of one matrix solver. My observation seems to be in line with your statement "Notice that most of the compute time is spent on the Jacobian operations" in the caption.
This behaviour of the solver, however, is counter-intuitive to me because the matrix solver should be the most expensive part in any implicit scheme. Please clarify why the formation of the Jacobian matrix is so much more expensive._

This is a very good observation and we also investigated this point further since the submission. Indeed the graphs presented in documentation (Solver performance) contain a mistake and will be rectified. The reported time for the Jacobian matrix evaluation and assembly was taken together with the preprocessing step of the MUMPS solver, which is  obviously a mistake. The preprocessing step in MUMPS corresponds to the factorization of the stiffness matrix of the linear system of equations. As to be expected, the factorization is the most memory and time consuming step for the direct solver. As mentioned in reply 4.), the computation and assembly of the Jacobian matrix is not the limiting factor regarding computational time, but will systematically require a new factorization within the direct solver which is computationally very expensive (roughly 2 orders of magnitude in comparison to the solving of the equations). There might be some options to interact with MUMPS directly through PETSc to control the re-factorization of the matrix for a given sparsity pattern. We have not investigated this part further at the moment.

**Comment by JaroslavHron** [comment]
Hi @albansouche , your turtleFSI is really nice tool to easily setup and run  FSI problems.

- I was missing some info on the used time discretization. In the setup you supply parameter `theta=0.5` which probably means Crank-Nicolson method, but in FSI example you use `theta=0.51` - ia it some way to stabilize the computation - and is it sill 2nd order? May be I just missed it or It can be found in the referenced theses.

- Running the provided benchmark setups, I have experienced one problem when running them in parallel - it seems that the point evaluation at his point [TF_fsi.py#L205](https://github.com/KVSlab/turtleFSI/blob/5c5c33e90a0f75a2048290182228a435e1295766/turtleFSI/problems/TF_fsi.py#L205) fails in parallel.
As far as I know it is a limitation in fenics, but there are some ways to overcome this limitation using mpi4py (see. 
https://fenicsproject.discourse.group/t/problem-with-evaluation-at-a-point-in-parallel)

- The only minor remark to the text I have: It might be useful to add the reference to the "Turek flag benchmarks" as is done in the documentation (https://turtlefsi2.readthedocs.io/en/latest/acknow_ref.html#id3) (but consider it as very minor remark - as I am coauthor of that publication and I am happy that it hase become such a popular test in part of the FSI community, however not everybody is familiar with that setup)

**Comment by albansouche** [comment]
Hi @JaroslavHron,
Thank you for the nice feedback and comments. Here are my replies: 

Comment 1: Yes, theta=1 corresponds to a fully implicit scheme, theta=0 to a fully explicit scheme, and theta=0.5 to a Crank-Nicolson scheme. One may also use the so-called shifted Crank-Nicolson scheme that corresponds to 0.5 + dt (the time step), “which allows for global stability of the solution” (Wick, 2011) and preserves the convergence properties of the Crank-Nicolson scheme. In the examples we use a shifted Crank-Nicolson scheme with value of 0.51 when dt is set to 0.01 s.
You can find the reference here: Wick, T., 2011. Adaptive finite element simulation of fluid-structure interaction with application to heart-valve dynamics (Doctoral dissertation). https://doi.org/10.11588/heidok.00012992

- Comment 2: Thank you for pointing this out and it is, indeed, a FEniCS limitation. The issue arises when evaluating the displacement field at a given mesh point in the _post_solve_ function of the problem file. The user can modify this file as they want without impacting turtleFSI solver. So, this is not related to turtleFSI solver. However, since it might be of interest for the future users to execute the solver in parallel, we now added the fix in the examples TF_fsi.py and TF_csm.py.

- Comment 3:  Yes, we now added the reference in the text as well.

**Comment by albansouche** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by albansouche** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by albansouche** [comment]
Hi @chennachaos, 
Following your comment on the computational cost of the different parts of the solver, I modified some of the text and the figure present in the documentation (https://turtlefsi2.readthedocs.io/en/latest/verif_perf.html).

**Comment by chennachaos** [comment]
Thanks for your response, @albansouche.

I have the following comments regarding the software paper.

* **Summary:** is nicely written.

* **A statement of need:** It is not clear what the motivation is and who the target audience are. Although the monolithic FSI solvers are robust, they are notorious to be computationally expensive beyond 100,000 DOFs, even for benchmark problems. Please note that benchmark problems are not always simple; one requires to run a considerable number of simulations (i) to assess the convergence of overall scheme, (ii) how different ingredients affect the accuracy, convergence and robustness, and (iii) to evaluate the parallel performance of the software tool.

* **State of the field:** This aspect is missing from the paper. There is no discussion and comparison with other FSI solvers (codes) available today. Oomph-lib and Elmer FEM are two such full-fledged open-source solvers. PreCICE is an open-source coupling library for FSI problems. Why should one prefer this FSI solver over others?

* **Quality of writing:** The paper is well written.

* **References:** The list of references needs to be improved by adding some classical references in the field of FSI.

**Comment by chennachaos** [comment]
I still have some issues in understanding the computational times reported in the "Solver performance" page.
In a tetrahedral mesh with 63,000 first-order elements, there would be about 15000 nodes. Based on this, I estimate the total DOFs for the monolithic scheme to be about 60,000. My experience indicates that the matrix factorisation step should not take 1000 seconds (~16 minutes) (assuming that the optimisation flags are turned ON) for a matrix of this size.
I tried to understand this further from the graph, but it is not clear to me what "jac.(1) + fact.(1) + it.(5)" represents. Does it mean that you performed 5 iterations by computing and factorising the Jacobian matrix once? If this the case, then (a) how does the error norm of the residual converge and (b) what is the tolerance for convergence criterion?

I would like to see the computational time split into three parts: (i) for computing the Jacobian, (ii) for the factorisation and (iii) for the solver, if you reuse the Jacobian matrix, for the first 100 time steps. Also, please do provide the convergence of the residual norm in Newton-Raphson iterations for a few random time steps in the first 100 time steps.

**Comment by aslakbergersen** [comment]
@chennachaos, thank you for your comments. We apologize for the delayed response. 

> Thanks for your response, @albansouche.
> 
> I have the following comments regarding the software paper.
> 
> * **Summary:** is nicely written.
> * **A statement of need:** It is not clear what the motivation is and who the target audience are. Although the monolithic FSI solvers are robust, they are notorious to be computationally expensive beyond 100,000 DOFs, even for benchmark problems. Please note that benchmark problems are not always simple; one requires to run a considerable number of simulations (i) to assess the convergence of overall scheme, (ii) how different ingredients affect the accuracy, convergence and robustness, and (iii) to evaluate the parallel performance of the software tool.
> * **State of the field:** This aspect is missing from the paper. There is no discussion and comparison with other FSI solvers (codes) available today. Oomph-lib and Elmer FEM are two such full-fledged open-source solvers. PreCICE is an open-source coupling library for FSI problems. Why should one prefer this FSI solver over others?
> * **Quality of writing:** The paper is well written.
> * **References:** The list of references needs to be improved by adding some classical references in the field of FSI.

To address the above points, we have added the text following below to the paper, with changes highlighted in bold. We have further specified the motivation, target audience, and why *some* could prefer turtleFSI over other solvers. We have also added some reference to other open-source solvers with FSI capabilities:

(…), but where physical insight might suffice. The aim was therefore to develop a fully monolithic and robust entry-level research code **with ease-of-use targeted towards students, educators, and researchers.**

FEniCS [@Logg:2012] has emerged as one (…) FEniCS was a natural choice of computing environment. **Compared to other open-source FSI solvers [@Malinen:2013; @Heil:2006; @Jasak:2007], turtleFSI is written in only a couple of hundred lines of high-level Python code, in contrast to tens of thousands of lines of low-level C++ code. This provides full transparency and a unique opportunity for researchers and educators to modify and experiment with the code, while still providing out of the box entry-level high-performance computing capabilities. Furthermore, because of the close resemblance between mathematics and code in FEniCS, users can make additions or modifications with ease.**

The turtleFSI solver rely on a fully monolithic (…) which can easily be adapted to other 2D or 3D FSI problems.

In conclusion, **turtleFSI is not a superior FSI solver in terms of speed, but it is a robust entry-level FSI solver** and performs exactly as designed and intended; ‘slow and steady wins the race’

----------------------------------------------

We have extended the section *Solver performance* from the documentation with a new section *Newton solver convergence and performance*. The documentation now includes a comparison of the Newton solver convergence between performing a full update and reusing the Jacobian matrix. This should answer the questions and concerns quoted below on the convergence behavior of the Newton solver used in different conditions. Of course, the reported behavior is not unique and depends on the geometry and boundary conditions but reflects well the general behavior we have observed using turtleFSI in our research work.

>I still have some issues in understanding the computational times reported in the "Solver performance" page. In a tetrahedral mesh with 63,000 first-order elements, there would be about 15000 nodes. Based on this, I estimate the total DOFs for the monolithic scheme to be about 60,000.

Our 3D FSI pipe flow problem has 460,000 dofs when discretized with 63000 tetrahedral elements. The pressure is approximated with a continuous linear scalar field, but the velocity and the deformation are both solved with a continuous quadratic vector field for stability reason.

>My experience indicates that the matrix factorisation step should not take 1000 seconds (~16 minutes) (assuming that the optimisation flags are turned ON) for a matrix of this size.
I tried to understand this further from the graph, but it is not clear to me what "jac.(1) + fact.(1) + it.(5)" represents. Does it mean that you performed 5 iterations by computing and factorising the Jacobian matrix once?

Yes, we compute and factorize the Jacobian matrix once at the first iteration and reused it for the next iterations. The factorization of the Jacobian matrix (with size 460000*460000) takes indeed ca. 1000 seconds running on 1 core and 1 thread.

> If this the case, then (a) how does the error norm of the residual converge and (b) what is the tolerance for convergence criterion? I would like to see the computational time split into three parts: (i) for computing the Jacobian, (ii) for the factorisation and (iii) for the solver, if you reuse the Jacobian matrix, for the first 100 time steps. Also, please do provide the convergence of the residual norm in Newton-Raphson iterations for a few random time steps in the first 100 time steps.

We have performed additional runs and documented the solver convergence rate in the new version of the section “Newton solver convergence and performance” of the software documentation.

**Comment by aslakbergersen** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1146/annurev.fluid.30.1.539 is OK
- 10.1023/A:1020843529530 is OK
- 10.1016/S0045-7825(00)00381-9 is OK
- 10.1016/j.compstruc.2011.02.019 is OK
- 10.1007/978-3-642-23099-8 is OK
- 10.1007/3-540-34596-5_15 is OK

MISSING DOIs

- https://doi.org/10.1007/3-540-34596-5_2 may be missing for title: oomph-lib–an object-oriented multi-physics finite-element library

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@albansouche @aslakbergersen can you check for that potentially missing DOI? thanks.

**Comment by aslakbergersen** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1146/annurev.fluid.30.1.539 is OK
- 10.1023/A:1020843529530 is OK
- 10.1016/S0045-7825(00)00381-9 is OK
- 10.1016/j.compstruc.2011.02.019 is OK
- 10.1007/978-3-642-23099-8 is OK
- 10.1007/3-540-34596-5_15 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1007/3-540-34596-5_2 is INVALID because of 'https://doi.org/' prefix
```

**Comment by aslakbergersen** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1146/annurev.fluid.30.1.539 is OK
- 10.1023/A:1020843529530 is OK
- 10.1016/S0045-7825(00)00381-9 is OK
- 10.1016/j.compstruc.2011.02.019 is OK
- 10.1007/978-3-642-23099-8 is OK
- 10.1007/3-540-34596-5_15 is OK
- 10.1007/3-540-34596-5_2 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@nathanquinlan, @chennachaos, @JaroslavHron thanks for your review efforts! We greatly appreciate your time and help! 

@chennachaos It looks like the authors made a number of changes which look ready for you to review. 

@JaroslavHron your last unticked box relates to `references`, I believe these have now been fixed. Can you check and if possible tick the last box? Thanks!

@nathanquinlan are you able to start with your part of the review process? Thanks!

**Comment by aslakbergersen** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by chennachaos** [comment]
Hi @aslakbergersen 
Thank you for your detailed response and for your updates to the paper and the documentation.

**Comment by chennachaos** [comment]
I have checked the installation process and found that the installation is quite involved especially if one does not have **anaconda** installed. (Conda requires about 4-5 GB disk space).

However, I have used the development version and installed it by following the instructions in the documentation page. Everything worked without any issues. I was also able to run the default test case; I had to kill the job though, as the default settings (3000 time steps) requires a lot of time to run. My suggestion is to modify the default settings for the test cases so that the overall simulation time for the test cases is limited to a few minutes, perhaps with an accompanying note to the user.

It is fantastic that you included the well-known benchmark FSI example from the Turek and Hron paper. However, test cases with the time-periodic response of the structure require a significant amount of time to get some meaningful results beyond the onset of vortex-shedding.
To make it easy for the novice users to understand the code and run test cases, I suggest including a numerical example with a steady-state response, for example, a beam in crossflow, as studied in https://www.sciencedirect.com/science/article/pii/S0045782518301026

**Comment by aslakbergersen** [comment]
Hi @chennachaos, I apologize for the delay. See below for answers to your comments.
 
> I have checked the installation process and found that the installation is quite involved especially if one does not have anaconda installed. (Conda requires about 4-5 GB disk space).

Thank you for your diligence. We have added a note that recommends users to install miniconda over the full anaconda, and provide the following [link](https://docs.conda.io/projects/conda/en/latest/user-guide/install/download.html#anaconda-or-miniconda) for users to learn more about the difference.

> However, I have used the development version and installed it by following the instructions in the documentation page. Everything worked without any issues. I was also able to run the default test case; I had to kill the job though, as the default settings (3000 time steps) requires a lot of time to run. My suggestion is to modify the default settings for the test cases so that the overall simulation time for the test cases is limited to a few minutes, perhaps with an accompanying note to the user.
>
> It is fantastic that you included the well-known benchmark FSI example from the Turek and Hron paper. However, test cases with the time-periodic response of the structure require a significant amount of time to get some meaningful results beyond the onset of vortex-shedding.
To make it easy for the novice users to understand the code and run test cases, I suggest including a numerical example with a steady-state response, for example, a beam in crossflow, as studied in https://www.sciencedirect.com/science/article/pii/S0045782518301026

This is an excellent point we as developers did not think about. The `turtle_demo.py` setup was initially designed to simulate a turtle swimming over many cycles to check the stability of the setup and create the animation for the repository. We have now reduced the total time of the simulation of the turtle demo problem to execute only the first 30 time steps, which should run in a couple of minutes on a desktop computer. We have also added a note in the documentation to explain this further.

**Comment by chennachaos** [comment]
Hi @aslakbergersen. Thank you for response and for the changes you made to the code and the documentation.

**Comment by chennachaos** [comment]
HI @Kevin-Mattheus-Moerman. I now have completed my review.

The authors have addressed all the issues and improved the quality of code, tests and documentation to a satisfactory level.
I recommend the **turtleFSI** library for publication in the JOSS.

Thank you very much for the opportunity!

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @chennachaos for your help here!!!!

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by nathanquinlan** [comment]
Dear all,
Sorry I haven't been able to contribute to the review process to date.  I don't want to delay the process further so I have no objection if the editors would like to proceed without my review.
Regards
Nathan

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon remove @nathanquinlan as reviewer

**Comment by whedon** [comment]
OK, @nathanquinlan is no longer a reviewer

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1146/annurev.fluid.30.1.539 is OK
- 10.1023/A:1020843529530 is OK
- 10.1016/S0045-7825(00)00381-9 is OK
- 10.1016/j.compstruc.2011.02.019 is OK
- 10.1007/978-3-642-23099-8 is OK
- 10.1007/3-540-34596-5_15 is OK
- 10.1007/3-540-34596-5_2 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @chennachaos @JaroslavHron for your review efforts! We will now process this work for acceptance in JOSS

**Comment by Kevin-Mattheus-Moerman** [comment]
@albansouche I will process your paper for acceptance. 

Here are some comments on your paper: 
- [ ] Replace `rely` with `relies` in: `The turtleFSI solver rely on a fully monolithic approach...` 
- [ ] in this sentence `We implemented and evaluated four different mesh lifting operators, ranging from simple and efficient 2nd order Laplace equation, most suitable for small deformations, to more sophisticated and computationally expensive 4th order bi-harmonic equation that can handle larger mesh deformations.`, it looks like this needs either `a simple` and `a more` (to go with the singular `equation`) or `equation` should be plural as in `equations`, please check. 
- [ ] Minor typo in `...although the later is...`, this should have `latter`. 
- [ ] Remove the `Installation and Use` section, this should be in the documentation and should therefore not appear in the paper. 
- [ ] Add s to illustration in: `Here are some illustration of the...` i.e. use `illustrations`
- [ ] I would recommend increasing the size of the image in the paper, e.g. full width. 
- [ ] Your paper is about to be accepted in JOSS, please thoroughly review/reread the paper yourself and check for typos etc. In particular check that the author names, affiliations, and acknowledgements are accurate. 

Once you've made the above changes you can run `@whedon generate pdf` here to update the paper. 

Finally please also do the following: 
- [ ] Please archive a copy of the software on Zenodo and report back to us with the DOI of the archived version. Ensure that the archived version meta data on Zenodo matches the paper here. E.g. the author list and the title should match the JOSS paper. You can edit this after creating the archived version if needed. 
- [ ] Can you confirm what the version of the final reviewed and archived version is? Is it still `v1.3`?

**Comment by aslakbergersen** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by aslakbergersen** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02089/joss.02089/10.21105.joss.02089.pdf)

**Comment by aslakbergersen** [comment]
@Kevin-Mattheus-Moerman thank you for your comments. I have implemented all your comments and checked for typos.

> Please archive a copy of the software on Zenodo and report back to us with the DOI of the archived version. Ensure that the archived version meta data on Zenodo matches the paper here. E.g. the author list and the title should match the JOSS paper. You can edit this after creating the archived version if needed.
10.5281/zenodo.3895443

> [Can you confirm what the version of the final reviewed and archived version is? Is it still v1.3?
To include the changes in the review to the distributed version on conda-forge I bumped the version number to 1.4. Do I need to edit this somewhere else? The DOI refers to the v1.4

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set 10.5281/zenodo.3895443 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3895443" target="_blank">10.5281/zenodo.3895443</a> is the archive.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set v1.4 as version

**Comment by whedon** [comment]
OK. v1.4 is the version.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1146/annurev.fluid.30.1.539 is OK
- 10.1023/A:1020843529530 is OK
- 10.1016/S0045-7825(00)00381-9 is OK
- 10.1016/j.compstruc.2011.02.019 is OK
- 10.1007/978-3-642-23099-8 is OK
- 10.1007/3-540-34596-5_15 is OK
- 10.1007/3-540-34596-5_2 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1490

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1490, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1272851068339314696) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1491
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02089](https://doi.org/10.21105/joss.02089)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02089/status.svg)](https://doi.org/10.21105/joss.02089)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02089">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02089/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02089/status.svg
   :target: https://doi.org/10.21105/joss.02089
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02089/status.svg)](https://doi.org/10.21105/joss.02089)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
